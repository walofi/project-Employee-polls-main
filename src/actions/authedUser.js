export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const CHECK_AUTHED_USER = "CHECK_AUTHED_USER";

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}
