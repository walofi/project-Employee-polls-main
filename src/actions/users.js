export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_ANSWER = "ADD_ANSWER";
export const ADD_QUESTION_TO_USER = "ADD_QUESTION_TO_USER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addAnswer({ authedUser, qid, answer }) {
  return {
    type: ADD_ANSWER,
    authedUser,
    qid,
    answer,
  };
}
export function addQuestionToUser({ authedUser, qid }) {
  return {
    type: ADD_QUESTION_TO_USER,
    authedUser,
    qid,
  };
}
