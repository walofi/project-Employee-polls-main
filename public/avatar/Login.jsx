/* eslint-disable react/prop-types */
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import { useRef } from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";

function Login(props) {
  const navigate = useNavigate();
  const usernameRef = useRef("");
  const passwordRef = useRef("");

  const handleLogin = () => {
    const username = usernameRef.current.value;
    const password = passwordRef.current.value;
    if (username === "" || password === "") {
      alert("Please fill in all fields");
    } else if (Object.keys(props.users).includes(username)) {
      if (props.users[username].password === password) {
        localStorage.setItem("authedUser", username);
        props.dispatch(setAuthedUser(username));
        if (localStorage.getItem("prevPath") !== null) {
          window.location.replace(
            location.origin + localStorage.getItem("prevPath")
          );
          localStorage.removeItem("prevPath");
        } else navigate("/");
      } else {
        alert("Incorrect password or username");
      }
    } else {
      alert("Incorrect password or username");
    }
  };
  console.log(props.authedUser, localStorage.getItem("authedUser"));
  return (
    <Form>
      {props.authedUser !== undefined ||
      localStorage.getItem("authedUser") !== null ? (
        <div>
          <h2>You are already logged in.</h2>
          <p>If you want to log out, please click logout above.</p>
        </div>
      ) : (
        <>
          <h2>Login</h2>
          <hr></hr>
          <Form.Group className="mb-3" controlId="formBasicUsername">
            <Form.Label>Username</Form.Label>
            <Form.Control
              ref={usernameRef}
              type="username"
              placeholder="Enter username"
            />
          </Form.Group>

          <Form.Group className="mb-3" controlId="formBasicPassword">
            <Form.Label>Password</Form.Label>
            <Form.Control
              ref={passwordRef}
              type="password"
              placeholder="Password"
            />
          </Form.Group>

          <Button id="LoginBtn" onClick={handleLogin} variant="primary">
            Login
          </Button>
        </>
      )}
    </Form>
  );
}

const mapStateToProps = ({ users }) => {
  return {
    users,
  };
};

export default connect(mapStateToProps)(Login);
