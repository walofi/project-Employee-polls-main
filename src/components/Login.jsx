
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import { useRef } from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";

function Login(props) {
  const navigate = useNavigate();
  const usernameRef = useRef("");
  const passwordRef = useRef("");

  const handleLogin = () => {
    const username = usernameRef.current.value;
    const password = passwordRef.current.value;

    if (username === "" || password === "") {
      alert("Please fill in all fields");
    } else if (Object.keys(props.users).includes(username)) {
      if (props.users[username].password === password) {
        localStorage.setItem("authedUser", username);
        props.dispatch(setAuthedUser(username));

        if (localStorage.getItem("prevPath") !== null) {
          // Use history.push for React Router navigation
          navigate(localStorage.getItem("prevPath"));
          localStorage.removeItem("prevPath");
        } else {
          navigate("/");
        }
      } else {
        alert("Incorrect password or username");
      }
    } else {
      alert("Incorrect password or username");
    }

    usernameRef.current.value = "";
    passwordRef.current.value = "";
  };
  return (
    <Form>
      {props.authedUser !== undefined ||
      localStorage.getItem("authedUser") !== null ? (
        <div>
          <h2>You are already logged in.</h2>
          <p>If you want to log out, please click logout above.</p>
        </div>
      ) : (
        <>
          <h2>Login</h2>
          <hr></hr>
          <Form.Group className="mb-3" controlId="formBasicUsername">
            <Form.Label>Username</Form.Label>
            <Form.Control
              data-testid="username"
              ref={usernameRef}
              type="username"
              placeholder="Enter username"
            />
          </Form.Group>

          <Form.Group className="mb-3" controlId="formBasicPassword">
            <Form.Label>Password</Form.Label>
            <Form.Control
              data-testid="password"
              ref={passwordRef}
              type="password"
              placeholder="Password"
            />
          </Form.Group>

          <Button
            data-testid="LoginBtn"
            onClick={handleLogin}
            variant="primary"
          >
            Login
          </Button>
        </>
      )}
    </Form>
  );
}

const mapStateToProps = ({ users }) => {
  return {
    users,
  };
};

export default connect(mapStateToProps)(Login);
