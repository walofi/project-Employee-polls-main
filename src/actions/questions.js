import { _saveQuestionAnswer, _saveQuestion } from "../../_DATA";
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { addAnswer, addQuestionToUser } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddQuestion(question) {
  return (dispatch) => {
    dispatch(showLoading());

    return _saveQuestion(question)
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(
          addQuestionToUser({ qid: question.id, authedUser: question.author })
        );
      })
      .then(() => {
        dispatch(hideLoading());
        localStorage.removeItem("handleAnswer");
      });
  };
}

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    qid,
    authedUser,
    answer,
  };
}

export function handleAnswerQuestion(info) {
  return (dispatch) => {
    dispatch(showLoading());
    return _saveQuestionAnswer(info)
      .then(() => {
        dispatch(answerQuestion(info));
        dispatch(addAnswer(info));
      })
      .then(() => {
        dispatch(hideLoading());
        localStorage.removeItem("handleAnswer");
      })
      .catch((e) => {
        console.warn("Error in handleAnswerQuestion: ", e);
        alert("There was an error answering the question. Try again.");
      })
      .catch((e) => {
        console.error("Error in dispatching answerQuestion: ", e);
      });
  };
}
