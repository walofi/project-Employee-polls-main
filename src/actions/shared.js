import { _getUsers, _getQuestions } from "../../_DATA";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";
import { setAuthedUser } from "./authedUser";
import { showLoading, hideLoading } from "react-redux-loading-bar";

const AUTHED_ID = localStorage.getItem("authedUser")
  ? localStorage.getItem("authedUser")
  : "";

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    _getUsers()
      .then((users) => {
        _getQuestions().then((questions) => {
          dispatch(receiveUsers(users));
          dispatch(receiveQuestions(questions));
          dispatch(setAuthedUser(AUTHED_ID));
          dispatch(hideLoading());
        });
      })
      .catch((error) => {
        alert(error);
      });
  };
}
