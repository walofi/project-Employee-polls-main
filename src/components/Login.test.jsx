import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../store";
import { BrowserRouter } from "react-router-dom";
import Login from "./Login";
import { handleInitialData } from "../actions/shared";
import "jest-localstorage-mock";

window.alert = jest.fn();

describe("Login", () => {
  it("should render the component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should clear input elements after clicking submit button", async () => {
    await store.dispatch(handleInitialData());

    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const usernameInputElement = screen.getByTestId("username");
    const passwordInputElement = screen.getByTestId("password");
    const loginButtonElement = screen.getByTestId("LoginBtn");

    expect(usernameInputElement.value).toBe("");
    expect(passwordInputElement.value).toBe("");

    fireEvent.change(usernameInputElement, { target: { value: "bh" } });
    fireEvent.change(passwordInputElement, {
      target: { value: "password123" },
    });

    expect(usernameInputElement.value).toBe("bh");
    expect(passwordInputElement.value).toBe("password123");

    fireEvent.click(loginButtonElement);

    expect(usernameInputElement.value).toBe("");
    expect(passwordInputElement.value).toBe("");
  });
});
