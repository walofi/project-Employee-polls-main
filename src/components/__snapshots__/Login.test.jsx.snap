// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <h2>
      Login
    </h2>
    <hr />
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicUsername"
      >
        Username
      </label>
      <input
        class="form-control"
        data-testid="username"
        id="formBasicUsername"
        placeholder="Enter username"
        type="username"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicPassword"
      >
        Password
      </label>
      <input
        class="form-control"
        data-testid="password"
        id="formBasicPassword"
        placeholder="Password"
        type="password"
      />
    </div>
    <button
      class="btn btn-primary"
      data-testid="LoginBtn"
      type="button"
    >
      Login
    </button>
  </form>
</DocumentFragment>
`;
